grub-core/loader/slaunch/x86_efi_linux.c

```diff
@@ -58,15 +58,51 @@ sl_efi_install_slr_table (struct grub_slaunch_params *slparams)
   return GRUB_ERR_NONE;
 }
 
+/* Traverses sections of a loaded EFI image to map an offset within a file to
+ * an offset within the image. Returns appropriately casted -1 on error. */
+static grub_efi_uint64_t
+foffset_to_voffset(grub_efi_loaded_image_t *image, grub_uint32_t foffset)
+{
+  struct grub_msdos_image_header *header;
+  struct grub_pe_image_header *pe_image_header;
+  struct grub_pe32_coff_header *coff_header;
+  struct grub_pe32_section_table *sections;
+  struct grub_pe32_section_table *section;
+  grub_uint16_t i;
+
+  header = image->image_base;
+  pe_image_header = (struct grub_pe_image_header *)
+      ((char *) header + header->pe_image_header_offset);
+  coff_header = &pe_image_header->coff_header;
+  sections = (struct grub_pe32_section_table *)
+      ((char *) coff_header + sizeof (*coff_header)
+       + coff_header->optional_header_size);
+
+  for (i = 0, section = sections; i < coff_header->num_sections; i++, section++)
+    {
+      grub_uint32_t fstart = section->raw_data_offset;
+      grub_uint32_t fend = fstart + section->raw_data_size;
+
+      if ((section->characteristics & 0x60) != 0 &&
```

Comment:
Please use `GRUB_PE32_SCN_CNT_CODE` and `GRUB_PE32_SCN_CNT_INITIALIZED_DATA` for better readability, and maybe a variable for it to keep the line length reasonable.

Reply:
0727266682c09416ecc96260ac3f445ac6785eb2